File: Subset Sums - GFG/subset-sums.java
//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t-- > 0)
        {
            int N=sc.nextInt();
			ArrayList<Integer> arr = new ArrayList<>();
			for(int i = 0; i < N ; i++){
			    arr.add(sc.nextInt());
			}
            Solution ob = new Solution();
         
            ArrayList<Integer> ans = ob.subsetSums(arr,N);
            Collections.sort(ans);
            for(int sum : ans){
                System.out.print(sum+" ");
            }
            System.out.println();
        }  
    }
}

// } Driver Code Ends


//User function Template for Java//User function Template for Java
class Solution{
    private void find(int ind,ArrayList<Integer>arr,int n,ArrayList<Integer>ans,int sum){
        if(ind==n){
            ans.add(sum);
            return;
        }
        //pick
        find(ind+1,arr,n,ans,sum+arr.get(ind));
        //not pick
        find(ind+1,arr,n,ans,sum);
    }
    ArrayList<Integer> subsetSums(ArrayList<Integer> arr, int n){
        // code here
        ArrayList<Integer> ans=new ArrayList<>();
        int s=0;
        find(0,arr,n,ans,s);
        Collections.sort(ans);
        return ans;
    }
}

File: Subset Sums - GFG/subset-sums.cpp
//{ Driver Code Starts
#include<bits/stdc++.h> 
using namespace std; 

// } Driver Code Ends
class Solution
{
public:
void p(int i,int s,vector<int>&arr,int n,vector<int>&ans){
    
    if(i==n){
        ans.push_back(s);
        return;
    }
    
    // ds.push_back(arr[i]);
    s+=arr[i];
    p(i+1,s,arr,n,ans);
    
    // ds.pop_back();
    s-=arr[i];
    p(i+1,s,arr,n,ans);
}
    vector<int> subsetSums(vector<int> arr, int n)
    {
        // Write Your Code here
        vector<int>ans;
        
        int s=0;
        p(0,s,arr,n,ans);
        return ans;
    }
};

//{ Driver Code Starts.
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int N;
        cin>>N;
        vector<int> arr(N);
        for(int i = 0 ; i < N ; i++){
            cin >> arr[i];
        }
        Solution ob;
        vector<int> ans = ob.subsetSums(arr,N);
        sort(ans.begin(),ans.end());
        for(auto sum : ans){
            cout<< sum<<" ";
        }
        cout<<endl;
    }
    return 0;
}
// } Driver Code Ends